lib360dataquality/cove/schema.py

```diff
@@ -1,13 +1,187 @@
+        return json.dumps(self._pkg_schema_obj)
+
+    def write_schema_file(self):
+        with open(self.schema_file, "w") as fp:
+            fp.write(self.schema_str)
+
+    def write_pkg_schema_file(self):
+        with open(self.pkg_schema_file, "w") as fp:
+            fp.write(self.pkg_schema_str)
+
+    def process_codelists(self):
+        # From libcove common but with support for codelists from 360Giving extensions added.
+
+        self.core_codelist_schema_paths = get_schema_codelist_paths(
+            self, use_extensions=False
+        )
+
+        extension_unique_files = frozenset(
+            url.split("/")[-1] for url in self.extension_codelist_urls
+        )
+
+        core_unique_files = frozenset(
+            value[0] for value in self.core_codelist_schema_paths.values() if value[0] not in extension_unique_files
+        )
+
+        # This loader uses the codelist host from the config + filename that was taken out of the schema
+        self.core_codelists = load_core_codelists(
+            self.codelists,
+            core_unique_files,
+            config=self.config if hasattr(self, "config") else None,
+        )
+
+        extension_codelists = {}
+
+        for extension_codelist_url in self.extension_codelist_urls:
+            codelist_file = extension_codelist_url.split("/")[-1]
+
+            extension_codelists[codelist_file] = load_codelist(
+                extension_codelist_url,
+                config=self.config if hasattr(self, "config") else None)
+
+        # Update the codelists with any specified by the extension
+        # This has the unfortunate side-effect of making cove think these are part of
+        # the main standard however we have no current way to differentiate the paths
+        self.core_codelists.update(extension_codelists)
+
+        # Ignore. Properties required by libcove:
+        self.extended_codelist_schema_paths = self.core_codelist_schema_paths
+        self.extended_codelists = self.core_codelists
+        self.extended_codelist_urls = {}
+        # End ignore
+
+        # we do not want to cache if the requests failed.
+        if not self.core_codelists:
+            load_core_codelists.cache_clear()
+            return
+
+    def resolve_extension(self, json_data) -> Optional[list]:
+        """
+        If json_data contains an extension id this patches the schemas if the extension is valid
+        the internal representation of the schema is replaced with the new patched version.
+        We write the new schema file(s) to disk for flattentool and caching purposes.
+
+        Returns an array of extension_infos or None
+        """
+
+        # FIXME NOTE this currently requires env SCHEMA_BRANCH=1.4-staging
+
+        try:
+            extension_ids = json_data["extensions"]
+        except KeyError:
+            return None
+
+        if len(extension_ids) == 0:
+            raise ExtensionsError("Extension key found but with no value(s)")
```

Comment:
This is so that it bubbles up to the front end, exceptions are shown via a decorator on the view capturing certain exceptions. https://github.com/OpenDataServices/lib-cove-web/blob/main/cove/views.py#L20